n4js-libs/packages/n4js-cli/src/n4js/index.n4js

```diff
@@ -73,118 +68,233 @@ export public function n4jsc(
+            log("picked up property 'n4jsc.jar' from config file " + pathN4jscliConfigFile);
+            return customN4jscName;
+        }
     }
+    // use default/shipped jar
+    return null;
+}
 
-    const vmOptionsFieldNames = new Set<string>(VmOptions.n4type.dataFields().map(f => f.name));
-    for (const [key, value] of Object.entries(vmOptions)) {
-        if (!vmOptionsFieldNames.has(key)) {
-            log("unknown vm option: key=" + key + " value=" + value);
-            continue;
+function getN4jscjarFromConfig(pathN4jscliConfigFile : string) : string {
+    if (pathN4jscliConfigFile) {
+        const data = fs.readFileSync(pathN4jscliConfigFile, {encoding : 'utf8'});
+        const n4jscliConfig : any+ = JSON.parse(data);
+        if (n4jscliConfig && n4jscliConfig["n4jsc.jar"]) {
+            return n4jscliConfig["n4jsc.jar"] as string;
         }
-        const isBool = typeof value === "boolean";
-        if (isBool && !value) {
-            continue; // skip false boolean flags
+    }
+    return null;
+}
+
+export public function findN4jscliConfig(workingDir : string) : string {
+    let dir = path.resolve(workingDir);
+    while (true) {
+        if (dir.length < 2) {
+            return null;
         }
-        if (value) {
-            const option = capitalizeFirstLetter(key) + (isBool? '': value);
-            args.push("-" + option);
+        const pathN4jscliConfig = path.resolve(dir, N4JS_CLI_CONFIG);
+        if (fs.existsSync(pathN4jscliConfig)) {
+            return pathN4jscliConfig;
         }
+
+        dir = path.dirname(dir);
     }
 
-    return args;
+    return null;
 }
 
-function getN4jscArgs(goal: Goal, dir: string, options: ~Options+, log: {function(string)}): string[] {
-    const args: string[] = [];
-    if (goal) {
-        args.push(goal);
-    }
-    if (dir) {
-        args.push(dir);
+/**
+ * Returns the file name of the given custom n4jsc.jar property.
+ * Also ensures the presence of the given jar.
+ */
+async function ensureCustomN4jscJar(customN4jscName: string, log: {function(string)}) : string {
+    if (fs.existsSync(customN4jscName)) {
+        // a local file
+        return customN4jscName;
     }
 
-    if (options) {
-        const optionsFieldNames = new Set<string>(Options.n4type.dataFields().map(f => f.name));
-        for (const [key, value] of Object.entries(options)) {
-            if (!optionsFieldNames.has(key)) {
-                log("unknown option: key=" + key + " value=" + value);
-                continue;
-            }
-            const isBool = typeof value === "boolean";
-            if (isBool && !value) {
-                continue; // skip false boolean flags
-            }
-            args.push((key.length > 2 ? "--" : "-") + key);
-            if (!isBool) {
-                args.push(String(value));
+    const fileName = path.basename(customN4jscName);
+    const jarInBinDir = path.resolve(BIN_DIR, fileName);
+    if (fs.existsSync(jarInBinDir)) {
```

Comment:
With this check whether a jar was already downloaded, all URIs will be "ignored" that have a last segment equal to the file name of any previously downloaded jar. In particular, this applies to all URIs with a last segment equal to the default name "n4jsc.jar". This might be confusing for the user.